File: Go/electric-rose.go
package main

import (
	"fmt"
	"math"
)

type Battery struct {
	brand  string
	dry    bool
	charge float64
}

func (battery Battery) Print() string {
	return fmt.Sprintf("%s, %t, %.1f\n", battery.brand, battery.dry, battery.charge)
}

func UpdateCharge(batteries []*Battery) {
	for i := 0; i < len(batteries); i++ {

		if batteries[i].charge >= 0.1 {
			batteries[i].charge = math.Max(0, batteries[i].charge-0.1)
		}

		if batteries[i].dry {
			if batteries[i].charge >= 0.1 {
				batteries[i].charge = math.Max(0, batteries[i].charge-0.1)
			}
		}

	}
}

File: TypeScript-Playwright/app/electric-rose.ts
export class Battery {
    type: string;
    dry: boolean;
    charge: number;

    constructor(type: string, dry: boolean, charge: number) {
        this.type = type;
        this.dry = dry;
        this.charge = charge;
    }
}

export class ElectricRose {
    batteries: Battery[];

    constructor(batteries: Battery[] = []) {
        this.batteries = batteries;
    }

    updateCharge() {
        for (let i = 0; i < this.batteries.length; i++) {
            if (this.batteries[i].charge >= 0.1) {
                this.batteries[i].charge = Math.max(0, this.batteries[i].charge - 0.1);
            }

            if (this.batteries[i].dry) {
                if (this.batteries[i].charge >= 0.1) {
                    this.batteries[i].charge = Math.max(0, this.batteries[i].charge - 0.1);
                }
            }
        }
    }
}